clickhouse/base/sslsocket.cpp

```diff
@@ -0,0 +1,211 @@
+        HANDLE_SSL_CTX_ERROR(SSL_CTX_set_default_verify_paths(ctx.get()));
+    if (!context_params.path_to_ca_directory.empty())
+        HANDLE_SSL_CTX_ERROR(
+            SSL_CTX_load_verify_locations(
+                ctx.get(),
+                nullptr,
+                context_params.path_to_ca_directory.c_str())
+        );
+
+    for (const auto & f : context_params.path_to_ca_files)
+        HANDLE_SSL_CTX_ERROR(SSL_CTX_load_verify_locations(ctx.get(), f.c_str(), nullptr));
+
+    if (context_params.context_options != -1)
+        SSL_CTX_set_options(ctx.get(), context_params.context_options);
+    if (context_params.min_protocol_version != -1)
+        HANDLE_SSL_CTX_ERROR(
+            SSL_CTX_set_min_proto_version(ctx.get(), context_params.min_protocol_version));
+    if (context_params.max_protocol_version != -1)
+        HANDLE_SSL_CTX_ERROR(
+            SSL_CTX_set_max_proto_version(ctx.get(), context_params.max_protocol_version));
+
+    return ctx.release();
+}
+
+
+
+}
+
+#define HANDLE_SSL_ERROR(statement) do { \
+    if (const auto ret_code = statement; ret_code <= 0) \
+        throwSSLError(ssl_, SSL_get_error(ssl_, ret_code), LOCATION, STRINGIFY(statement)); \
+} while(false);
+
+namespace clickhouse {
+
+SSLContext::SSLContext(SSL_CTX & context)
+    : context_(&context)
+{
+    SSL_CTX_up_ref(context_);
+}
+
+SSLContext::SSLContext(const SSLParams & context_params)
+    : context_(prepareSSLContext(context_params))
+{
+}
+
+SSLContext::~SSLContext() {
+    SSL_CTX_free(context_);
+}
+
+SSL_CTX * SSLContext::getContext() {
+    return context_;
+}
+
+/* // debug macro for tracing SSL state
+#define LOG_SSL_STATE() std::cerr << "!!!!" << LOCATION << " @" << __FUNCTION__ \
+    << "\t" << SSL_get_version(ssl_) << " state: "  << SSL_state_string_long(ssl_) \
+    << "\n\t handshake state: " << SSL_get_state(ssl_) \
+    << std::endl
+*/
+SSLSocket::SSLSocket(const NetworkAddress& addr, const SSLParams & ssl_params, SSLContext& context)
+    : Socket(addr),
+    ssl_(SSL_new(context.getContext()))
+{
+    if (!ssl_)
+        throw std::runtime_error("Failed to create SSL instance");
+
+    HANDLE_SSL_ERROR(SSL_set_fd(ssl_, handle_));
+    if (ssl_params.use_SNI)
+        HANDLE_SSL_ERROR(SSL_set_tlsext_host_name(ssl_, addr.Host().c_str()));
+
+    SSL_set_connect_state(ssl_);
+    HANDLE_SSL_ERROR(SSL_connect(ssl_));
+    HANDLE_SSL_ERROR(SSL_set_mode(ssl_, SSL_MODE_AUTO_RETRY));
+
+    if(const auto verify_result = SSL_get_verify_result(ssl_); verify_result != X509_V_OK) {
+        auto error_message = X509_verify_cert_error_string(verify_result);
+        auto ssl_session = SSL_get_session(ssl_);
+        auto cert = SSL_SESSION_get0_peer(ssl_session);
+
+        throw std::runtime_error("Failed to verify SSL connection, X509_v error: "
+                                 + std::to_string(verify_result)
+                                 + " " + error_message + "\n" + getCertificateInfo(cert));
+    }
```

Comment:
Are you sure this is the right way to verify the certs of the peer?

> If no peer certificate was presented, the returned result code is X509_V_OK. This is because no verification error occurred, it does however not indicate success. SSL_get_verify_result() is only useful in connection with SSL_get_peer_certificate(3).

Reply:
good point, will also check if server actually provided any certificate